tests/integration/test_transpose.py

```diff
@@ -17,34 +17,194 @@
 
 import cunumeric as num
 
-rect = num.array([[1, 2, 3], [4, 5, 6]])
-square = num.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
-
-
-@pytest.mark.parametrize("x", (rect, square), ids=("rect", "square"))
-class Test_free_function:
-    def test_forward(self, x):
-        y = num.transpose(x)
-        npx = np.array(x)
-        assert num.array_equal(y, np.transpose(npx))
-
-    def test_round_trip(self, x):
-        y = num.transpose(x)
-        z = num.transpose(y)
-        assert num.array_equal(x, z)
-
-
-@pytest.mark.parametrize("x", (rect, square), ids=("rect", "square"))
-class Test_method:
-    def test_forward(self, x):
-        y = x.transpose()
-        npx = np.array(x)
-        assert num.array_equal(y, npx.transpose())
-
-    def test_round_trip(self, x):
-        y = x.transpose()
-        z = y.transpose()
-        assert num.array_equal(x, z)
+DIM = 5
+SIZES = [
+    (0,),
+    (1),
+    (DIM),
+    (1, 1),
+    (1, DIM),
+    (DIM, 1),
+    (DIM, DIM),
+    (1, 1, 1),
+    (DIM, 1, 1),
+    (1, DIM, 1),
+    (1, DIM - 1, DIM),
+    (2, DIM - 1, DIM),
+    (DIM, DIM, DIM),
+]
+
+
+@pytest.mark.xfail
+def test_none_array_compare():
+    res_num = num.transpose(None)  # AttributeError: 'NoneType'
+    res_np = np.transpose(None)  # return None
+    assert np.array_equal(res_num, res_np, equal_nan=True)
+
+
+def test_none_array():
+    # numpy returned None
+    msg = r"NoneType"
+    with pytest.raises(AttributeError, match=msg):
+        num.transpose(None)
+
+
+@pytest.mark.parametrize("axes", ((1, 1, 1), (1, 2, 3), (1, 2), (1, 2, 0, 1)))
+def test_num_invalid_axis(axes):
+    size = (2, 3, 4)
+    a = num.random.randint(low=-10, high=10, size=size)
+    with pytest.raises(ValueError):
+        num.transpose(a, axes=axes)
+
+
+def test_num_int_axis():
+    size = (2, 3, 4)
+    a = num.random.randint(low=-10, high=10, size=size)
+    # numpy raises "ValueError: axes don't match array"
+    # cunumeric raises "TypeError"
+    with pytest.raises(TypeError):
+        num.transpose(a, axes=2)
+
+
+@pytest.mark.parametrize("size", SIZES, ids=str)
+def test_num_basic(size):
+    a = np.random.randint(low=-10, high=10, size=size)
+    res_np = np.transpose(a)
+    res_num = num.transpose(a)
+    assert np.array_equal(res_num, res_np)
+
+
+@pytest.mark.parametrize("size", SIZES, ids=str)
+def test_array_basic(size):
+    a = np.random.randint(low=-10, high=10, size=size)
+    b = num.array(a)
+    res_np = a.transpose()
+    res_num = b.transpose()
+    assert np.array_equal(res_num, res_np)
```

Comment:
What do you think about using separate classes to group and organize the tests?
```python
class TestModule:

    @pytest.mark.parametrize("size", SIZES, ids=str)
    def test_basic(self, size):
        ...

    def test_round(self, size):
        ...

     # all other "module" tests

class TestArrayMethod:

    @pytest.mark.parametrize("size", SIZES, ids=str)
    def test_basic(self, size):
        ...

    def test_round(self, size):
        ...  

    # all other "array" tests
```
This way it would not be necessary to encode "num" and "array" in every test name (but then individual tests would be "split up" so there is some trade-off as well)

Reply:
That's a good idea.